### PR TITLE
Iterable ordered map alternative with improved performance

### LINKED ISSUE #1151: OrderedMap interface channel performance overhead
### Describe the issue 
#763 add ordered map support, I use it in production env, found this interface have channel overhead.
image
### Steps to reproduce
benchmark result:
```
$ CLICKHOUSE_USE_DOCKER=false CLICKHOUSE_PASSWORD=xx go test . -bench OrderedMap -run OrderedMap
using random seed 1701524440354929000 for native tests
goos: darwin
goarch: arm64
pkg: github.com/ClickHouse/clickhouse-go/v2/tests
BenchmarkOrderedMapUseChanGo-10                   633183              1890 ns/op             120 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseChanNoGo-10            2601765               469.9 ns/op           256 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseSlice-10              11817420               100.6 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/ClickHouse/clickhouse-go/v2/tests    5.265s
```
https://github.com/hanjm/clickhouse-go/blob/0313d19d47520862538f565612bd79ef1890e7aa/tests/map_test.go
```go
func (om *OrderedMap) Keys() <-chan any {
	ch := make(chan any)
	go func() {
		defer close(ch)
		for _, key := range om.keys {
			ch <- key
		}
	}()
	return ch
}
func (om *OrderedMap) KeysUseChanNoGo() <-chan any {
	ch := make(chan any, len(om.keys))
	for _, key := range om.keys {
		ch <- key
	}
	close(ch)
	return ch
}

func (om *OrderedMap) KeysUseSlice() []any {
	return om.keys
}

func BenchmarkOrderedMapUseChanGo(b *testing.B) {
	m := NewOrderedMap()
	for i := 0; i < 10; i++ {
		m.Put(i, i)
	}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		for key := range m.Keys() {
			_, _ = m.Get(key)
		}
	}
}

func BenchmarkOrderedMapKeysUseChanNoGo(b *testing.B) {
	m := NewOrderedMap()
	for i := 0; i < 10; i++ {
		m.Put(i, i)
	}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		for key := range m.KeysUseChanNoGo() {
			_, _ = m.Get(key)
		}
	}
}

func BenchmarkOrderedMapKeysUseSlice(b *testing.B) {
	m := NewOrderedMap()
	for i := 0; i < 10; i++ {
		m.Put(i, i)
	}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		for key := range m.KeysUseSlice() {
			_, _ = m.Get(key)
		}
	}
}
```

a proposal is add a v2 interface.
```
type OrderedMapV2 interface {
	Get(key any) (any, bool)
	Put(key any, value any)
	Keys() []any
}
```
#### Environment
* Client version: v2.16.0
* Language version: go 1.21
* OS: linux 
* Interface: ClickHouse API / database/sql compatible driver
#### ClickHouse server
* ClickHouse Server version: 
* ClickHouse Server non-default settings, if any:
* `CREATE TABLE` statements for tables involved:
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary

### PR DESCRIPTION
**Pull request by hanjm**
## Summary
fix #1151, introduce a v2 interface for type assert.

## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added

### CODE REVIEW

**Comment by hanjm** [comment]
> maybe use below interface?
> 
> ```
> type MapIter interface {
> 	Next() bool
> 	Key() any
> 	Value() any
> }
> 
> type OrderedMapV2 interface {
> 	Put(key any, value any)
> 	Range() MapIter
> }
> ```
> 
> so we can use more efficient way to store map, for example: two slices
> 
> ```
> type Pairs struct {
> 	Keys []any
> 	Values []any
> }
> ```

Looks great.
```
$ CLICKHOUSE_USE_DOCKER=false CLICKHOUSE_PASSWORD=tps go test -gcflags="-l" . -bench OrderedMap -run OrderedMap
using random seed 1702120904860719000 for native tests
goos: darwin
goarch: arm64
pkg: github.com/ClickHouse/clickhouse-go/v2/tests
BenchmarkOrderedMapUseChanGo-10                   584689              2023 ns/op             120 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseChanNoGo-10            2493792               502.1 ns/op           256 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseSlice-10              11242239               123.8 ns/op             0 B/op          0 allocs/op
BenchmarkOrderedMapKeysUseIter-10               61061322                19.83 ns/op           16 B/op          1 allocs/op
PASS
ok      github.com/ClickHouse/clickhouse-go/v2/tests    7.274s
```

**Comment by timmy21** [comment]
you map_test.go has a bug
```
func (om *OrderedMap) Put(key any, value any) {
	if _, present := om.values[key]; present {
		om.values[key] = value
		return
	}
	om.keys = append(om.keys, key)
	om.values[key] = value
	om.valuesIter = append(om.valuesIter, value)
}

func (i *mapIter) Value() any {
	return i.om.valuesIter[i.iterIndex]
}
```
and, you can also benchmark reflect map
```
func BenchmarkReflectMapIter(b *testing.B) {
	m := make(map[int]int)
	for i := 0; i < 10; i++ {
		m[i] = i
	}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		value := reflect.Indirect(reflect.ValueOf(m))
		iter := value.MapRange()
		for iter.Next() {
			_ = iter.Key()
			_ = iter.Value()
		}
	}
}
```

File: tests/map_test.go [hunk 1]
```diff
@@ -354,3 +358,102 @@ func TestMapValuer(t *testing.T) {
...
+}
+
+type mapIter struct {
+	om        *OrderedMap
+	iterIndex int
+}
+
+func (i *mapIter) Next() bool {
+	i.iterIndex++
+	return i.iterIndex >= len(i.om.keys)
```
Comment (timmy21) [inline tests/map_test.go:392]: return i.iterIndex < len(i.om.keys)

File: tests/map_test.go [hunk 2]
```diff
@@ -354,3 +358,102 @@ func TestMapValuer(t *testing.T) {
...
+}
+
+func (i *mapIter) Key() any {
+	return i.om.keys[i.iterIndex]
+}
+
+func (i *mapIter) Value() any {
+	return i.om.values[i.iterIndex]
```
Comment (timmy21) [inline tests/map_test.go:400]: return i.om.valuesIter[i.iterIndex]

**Comment by hanjm** [comment]
add benchmark reflect map.
```
$ CLICKHOUSE_USE_DOCKER=false CLICKHOUSE_PASSWORD=tps go test -gcflags="-l" . -bench OrderedMap -run OrderedMap
using random seed 1702126424636270000 for native tests
goos: darwin
goarch: arm64
pkg: github.com/ClickHouse/clickhouse-go/v2/tests
BenchmarkOrderedMapUseChanGo-10                   600756              1970 ns/op             120 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseChanNoGo-10            2467634               483.4 ns/op           256 B/op          2 allocs/op
BenchmarkOrderedMapKeysUseSlice-10              11058926               110.1 ns/op             0 B/op          0 allocs/op
BenchmarkOrderedMapKeysUseIter-10               20183755                57.78 ns/op           16 B/op          1 allocs/op
BenchmarkOrderedMapReflectMapIter-10             1732650               697.0 ns/op           448 B/op         21 allocs/op
PASS
ok      github.com/ClickHouse/clickhouse-go/v2/tests    8.459s
```

File: tests/map_test.go (hunk 1 above)
Comment (hanjm) [reply tests/map_test.go]: fixed.

File: tests/map_test.go (hunk 2 above)
Comment (hanjm) [reply tests/map_test.go]: fixed.

File: tests/map_test.go [hunk 3]
```diff
@@ -236,10 +238,12 @@ func (om *OrderedMap) Get(key any) (any, bool) {
 func (om *OrderedMap) Put(key any, value any) {
 	if _, present := om.values[key]; present {
 		om.values[key] = value
+		om.valuesIter = append(om.valuesIter, value)
```
[the remark this answers is not in the document]
Comment (hanjm) [reply tests/map_test.go:242]: no, om.valuesIter it use for mapIter.
```
func (i *mapIter) Value() any {
	return i.om.valuesIter[i.iterIndex]
}
```

[the remark this answers is not in the document]
Comment (timmy21) [reply tests/map_test.go:242]: the key is already exists，ignore it is ok

[the remark this answers is not in the document]
Comment (hanjm) [reply tests/map_test.go:242]: ok

**Comment by jkaflik** [comment]
@hanjm thanks for your contribution! Before I jump into review, could you elaborate on this implementation please?

**Comment by timmy21** [comment]
> @hanjm thanks for your contribution! Before I jump into review, could you elaborate on this implementation please?

The problem I ran into was that when writing to map columns, a lot of time was wasted in the reflect operation
Screen Shot 2023-12-12 at 12 29 23 PM

**Comment by jkaflik** [comment]
I understand the reasoning, but I want to understand what is your proposal to solve this.

**Comment by timmy21** [comment]
> @timmy21 I understand the reasoning, but I want to understand what is your proposal to solve this.

By defining a map interface, this allows user to store map using different data structure, so that avoiding the reflect overhead. for example use Pairs:
```
type Pairs struct {
	Keys []any
	Values []any
}
```

File: lib/column/map.go [hunk 4]
```diff
@@ -42,6 +43,17 @@ type OrderedMap interface {
 	Keys() <-chan any
 }
 
+type MapIter interface {
+	Next() bool
+	Key() any
+	Value() any
+}
```
Comment (jkaflik) [inline lib/column/map.go:50]:
```suggestion
type MapIterator interface {
	Next() bool
	Key() any
	Value() any
}
```

File: lib/column/map.go [hunk 5]
```diff
@@ -42,6 +43,17 @@ type OrderedMap interface {
...
+
+type OrderedMapV2 interface {
+	Put(key any, value any)
+	Iter() MapIter
+}
+
```
Comment (jkaflik) [inline lib/column/map.go:56]:
```suggestion
type IterableOrderedMap interface {
	Put(key any, value any)
	Iterator() MapIterator
}
```

**Review by jkaflik — COMMENTED** [review]
Overall it looks good. Only minor naming convention changes. 

Would be great if we provide additional example.

**Comment by hanjm** [comment]
> Overall it looks good. Only minor naming convention changes.
> 
> Would be great if we provide additional example.

Thanks, done.

**Review by jkaflik — APPROVED** [review]

**Pull request merged**